pkg/auth/auth.go

```diff
@@ -0,0 +1,84 @@
+/*This file is part of kuberpult.
+
+Kuberpult is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Kuberpult is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2021 freiheit.com*/
+package auth
+
+import (
+	"context"
+	"google.golang.org/grpc/metadata"
+	"net/http"
+)
+
+type ctxMarker struct{}
+
+var (
+	ctxMarkerKey = &ctxMarker{}
+	defaultUser  = &User{
+		Email: "local.user@freiheit.com",
+		Name:  "defaultUser",
+	}
+)
+
+// Extract takes the User from middleware.
+// It always returns a User
+func Extract(ctx context.Context) *User {
+	// check if User is in go Context
+	u, ok := ctx.Value(ctxMarkerKey).(*User)
+	if !ok {
+		// check if User is in Metadata
+		md, _ := metadata.FromIncomingContext(ctx)
+		u = &User{
+			Email: md.Get("author-email")[0],
+			Name:  md.Get("author-username")[0],
+		}
+	}
+	return u
+}
+
+// ToContext adds the User to the context for extraction later.
+// Returning the new context that has been created.
+func ToContext(ctx context.Context, u *User) context.Context {
+	if u == nil || u.Email == "" {
```

Comment:
Technically, since `ToContext` will only be used in the `ServeHTTP` function in this file I think the user check should be moved somewhere else? or be completely removed because the user is always given? or do we keep it as best practice?

Reply:
You can't really remove it from the package because the ctxMarkerKey is only available in this package. I don't think it's an issue to have it here